library/fluid/fluid_extensions/src/main/java/org/quiltmc/qsl/fluid/api/FlowableFluidExtensions.java

```diff
@@ -0,0 +1,372 @@
+			double xOffset = (random.nextDouble() * 2.0 - 1.0) * (double) splashing.getDimensions(splashing.getPose()).width;
+			double zOffset = (random.nextDouble() * 2.0 - 1.0) * (double) splashing.getDimensions(splashing.getPose()).width;
+			int yFloor = MathHelper.floor(splashing.getY());
+			ParticleEffect particle = getBubbleParticle(splashing, splashPos, random);
+			if (particle != null) {
+				splashing.world.addParticle(
+						particle,
+						splashing.getX() + xOffset,
+						yFloor + 1.0f,
+						splashing.getZ() + zOffset,
+						splashPos.x,
+						splashPos.y - random.nextDouble() * 0.2,
+						splashPos.z
+				);
+			}
+
+		}
+	}
+
+	default void onSplash(World world, Vec3d pos, Entity splashing, RandomGenerator random) {
+		Entity passenger = splashing.hasPassengers() && splashing.getPrimaryPassenger() != null ? splashing.getPrimaryPassenger() : splashing;
+		float volumeMultiplier = passenger == splashing ? 0.2f : 0.9f;
+		Vec3d velocity = passenger.getVelocity();
+		double volume = Math.min(
+				1.0f,
+				Math.sqrt(velocity.x * velocity.x * 0.2 + velocity.y * velocity.y + velocity.z * velocity.z * 0.2D) * volumeMultiplier
+		);
+
+		SoundEvent sound = volume < 0.25f ? getSplashSound(splashing, pos, random) : getHighSpeedSplashSound(splashing, pos, random);
+		if (sound != null) {
+			splashing.playSound(
+					sound,
+					(float) volume,
+					1.0f + (random.nextFloat() - random.nextFloat()) * 0.4f
+			);
+		}
+
+		spawnBubbleParticles(splashing, pos, random);
+		spawnSplashParticles(splashing, pos, random);
+
+		GameEvent splash = getSplashGameEvent(splashing, pos, random);
+		if (splash != null) {
+			splashing.emitGameEvent(splash);
+		}
+	}
+
+	/**
+	 * Here you can modify viscosity and speed based on enchantments.
+	 *
+	 * @return a float array where [0] holds updated horizontal viscosity, and [1] holds updated speed.
+	 */
+	default FluidEnchantmentHelper customEnchantmentEffects(Vec3d movementInput, LivingEntity entity, float horizontalViscosity, float speed) {
+		float depthStriderLevel = EnchantmentHelper.getDepthStrider(entity);
+		if (depthStriderLevel > 3.0f) {
+			depthStriderLevel = 3.0f;
+		}
+
+		if (!entity.isOnGround()) {
+			depthStriderLevel *= 0.5f;
+		}
+
+		if (depthStriderLevel > 0.0f) {
+			horizontalViscosity += (0.546f - horizontalViscosity) * depthStriderLevel / 3.0f;
+			speed += (entity.getMovementSpeed() - speed) * depthStriderLevel / 3.0f;
+		}
+		return new FluidEnchantmentHelper(horizontalViscosity, speed);
+	}
+
+	default void doDrownEffects(FluidState state, LivingEntity drowning, RandomGenerator random) {
+		boolean isPlayer = drowning instanceof PlayerEntity;
+		boolean invincible = isPlayer && ((PlayerEntity) drowning).getAbilities().invulnerable;
+		if (!drowning.canBreatheInWater() && !StatusEffectUtil.hasWaterBreathing(drowning) && !invincible) {
+			drowning.setAir(getNextAirSubmerged(drowning.getAir(), drowning, random));
+			// if out of air
+			if (drowning.getAir() == -20) {
```

Comment:
This `-20` seems like a magic value. Please document it as such.

Reply:
I meant putting it in a constant, not literally adding Javadoc to it (which doesn't work)!

Reply:
oh sorry :C